src/core/public/notifications/notifications_service.ts

```diff
@@ -31,27 +32,45 @@ interface Deps {
 export class NotificationsService {
   private readonly toasts: ToastsService;
 
-  private readonly toastsContainer: HTMLElement;
+  private readonly toastsContainer$: Subject<HTMLElement>;
+  private domElemSubscription?: Subscription;
+  private targetDomElement?: HTMLElement;
 
   constructor(private readonly params: Params) {
-    this.toastsContainer = document.createElement('div');
+    this.toastsContainer$ = new Subject<HTMLElement>();
     this.toasts = new ToastsService({
-      targetDomElement: this.toastsContainer,
+      targetDomElement$: this.toastsContainer$.asObservable(),
     });
   }
 
   public setup({ i18n }: Deps) {
-    this.params.targetDomElement.appendChild(this.toastsContainer);
+    this.domElemSubscription = this.params.targetDomElement$.subscribe({
+      next: targetDomElement => {
+        this.cleanupTargetDomElement();
+        this.targetDomElement = targetDomElement;
 
-    return {
-      toasts: this.toasts.setup({ i18n }),
-    };
+        const toastsContainer = document.createElement('div');
+        targetDomElement.appendChild(toastsContainer);
+        this.toastsContainer$.next(toastsContainer);
+      },
+    });
+
+    return { toasts: this.toasts.setup({ i18n }) };
   }
 
   public stop() {
     this.toasts.stop();
+    this.cleanupTargetDomElement();
```

Comment:
nit: You introduce two different patterns in this stop method. In one, you always execute a separate cleanup function which wraps the conditional cleanup. Immediately after that you just have a the conditional cleanup directly inside `stop`. Not a huge deal, but consistency in our patterns is useful as we add more and more code over time.

Reply:
I'm not sure I'm seeing the inconsistency here. Both pieces conditionally cleanup their DOM element and subscriptions if they exist. Only difference is the NotificationsService (which is a parent of Toasts) calls `toasts.stop` before cleaning up the parent element.

```ts
  // notifications_service.ts
  public stop() {
    // cleans up the child element first
    this.toasts.stop();
    // then cleanup containing element
    this.cleanupTargetDomElement();

    if (this.domElemSubscription) {
      this.domElemSubscription.unsubscribe();
    }
  }

  private cleanupTargetDomElement() {
    if (this.targetDomElement) {
      this.targetDomElement.textContent = '';
    }
  }
```

```ts
  // toasts_service.tsx
  public stop() {
    this.cleanupTargetDomElement();

    if (this.domElemSubscription) {
      this.domElemSubscription.unsubscribe();
    }
  }

  private cleanupTargetDomElement() {
    if (this.targetDomElement) {
      // only the ToastsService has an actual React component to unmount.
      unmountComponentAtNode(this.targetDomElement);
      this.targetDomElement.textContent = '';
    }
  }
```